File: Proj4a/ackerman/Ackerman.java
public class Ackerman {
    
    public static int A(int m, int n, int p) {
        if (p == 0) {
            return m + n;
        } else if (n == 0 && p == 1) {
            return 0;
        } else if (n == 0 && p == 2) {
            return 1;
        } else if (n == 0 && p > 2) {
            return m;
        } else if (n > 0 && p > 0) {
            return A(m, A(m, n - 1, p), p - 1);
        }
        return 0; 
    }
    
    public static void main(String[] args) {
        System.out.println(A(7, 5, 2)); 
    }
}

File: Proj4a/ackerman/Ackerman.pl

ackerman(M, N, 0, Result) :-
    Result is M + N.

ackerman(_, 0, 1, 0).

ackerman(_, 0, 2, 1).

ackerman(M, 0, P, Result) :-
    P > 2,
    Result is M.

ackerman(M, N, P, Result) :-
    N > 0,
    P > 0,
    N1 is N - 1,
    P1 is P - 1,
    ackerman(M, N1, P, Temp),
    ackerman(M, Temp, P1, Result).


